src/libaktualizr/storage/sqlstorage_base.cc

```diff
@@ -61,6 +63,59 @@ std::string SQLStorageBase::getTableSchemaFromDb(const std::string& tablename) {
   return schema.value() + ";";
 }
 
+bool SQLStorageBase::dbMigrateForward(int version_from) {
+  SQLite3Guard db = dbConnection();
+
+  if (!db.beginTransaction()) {
+    LOG_ERROR << "Can't start transaction: " << db.errmsg();
+    return false;
+  }
+
+  for (int32_t k = version_from + 1; k <= current_schema_version_; k++) {
+    auto result_code = db.exec(schema_migrations_.at(static_cast<size_t>(k)), nullptr, nullptr);
+    if (result_code != SQLITE_OK) {
+      LOG_ERROR << "Can't migrate db from version " << (k - 1) << " to version " << k << ": " << db.errmsg();
+      return false;
+    }
+    if (schema_rollback_migrations_.at(static_cast<uint32_t>(k)).empty()) {
+      continue;
+    }
+    auto statement = db.prepareStatement("INSERT OR REPLACE INTO rollback_migrations VALUES (?,?);", k,
+                                         schema_rollback_migrations_.at(static_cast<uint32_t>(k)));
+    if (statement.step() != SQLITE_DONE) {
+      LOG_ERROR << "Can't insert rollback migration script: " << db.errmsg();
+      return false;
+    }
+  }
+
+  db.commitTransaction();
+
+  return true;
+}
+
+bool SQLStorageBase::dbMigrateBackward(int version_from) {
+  SQLite3Guard db = dbConnection();
+  for (int ver = version_from; ver > current_schema_version_; --ver) {
+    auto statement = db.prepareStatement("SELECT migration FROM rollback_migrations WHERE version_from=?;", ver);
+    if (statement.step() != SQLITE_ROW) {
+      LOG_ERROR << "Can't extract migration script: " << db.errmsg();
+      return false;
+    }
+    auto migration = statement.get_result_col_str(0);
+    statement.step();
+    if (db.exec(*migration, nullptr, nullptr) != SQLITE_OK) {
+      LOG_ERROR << "Can't migrate db from version " << (ver) << " to version " << ver - 1 << ": " << db.errmsg();
+      return false;
+    }
+    if (db.exec(std::string("DELETE FROM rollback_migrations WHERE version_from=") + std::to_string(ver) + ";", nullptr,
+                nullptr) != SQLITE_OK) {
```

Comment:
`DELETE` probably won't hurt, but shouldn't be necessary if we add with `INSERT OR REPLACE`.

Reply:
Here is the original discussion: https://github.com/advancedtelematic/aktualizr/pull/1072#discussion_r251880431. Now I don't feel too strongly about it.

Reply:
Sorry, missed that conversation. And I agree, I don't feel strongly, I just wanted to leave the thought.